crates/types/src/net/cluster_state.rs

```diff
@@ -0,0 +1,188 @@
+use std::{
+    collections::{BTreeMap, HashMap},
+    hash::{Hash, Hasher},
+};
+
+use serde::{Deserialize, Serialize};
+use serde_with::{serde_as, FromInto};
+
+use crate::{
+    cluster::cluster_state::{AliveNode, PartitionProcessorStatus},
+    identifiers::PartitionId,
+    time::MillisSinceEpoch,
+    GenerationalNodeId, PlainNodeId,
+};
+
+use super::{define_rpc, TargetName};
+
+#[serde_as]
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+pub struct NodeData {
+    pub timestamp: MillisSinceEpoch,
+    pub generational_node_id: GenerationalNodeId,
+    #[serde_as(as = "serde_with::Seq<(FromInto<u16>, _)>")]
+    pub partitions: BTreeMap<PartitionId, PartitionProcessorStatus>,
+}
+
+impl From<NodeData> for AliveNode {
+    fn from(value: NodeData) -> Self {
+        Self {
+            generational_node_id: value.generational_node_id,
+            last_heartbeat_at: value.timestamp,
+            partitions: value.partitions,
+        }
+    }
+}
+impl From<&NodeData> for NodeHash {
+    fn from(value: &NodeData) -> Self {
+        Self {
+            timestamp: value.timestamp,
+            hash: value.hashed(),
+        }
+    }
```

Comment:
I know it's not very common but there are some use case for it, otherwise it would have been enforced with boundaries. Anyway, I will replace it and try to avoid this pattern in the future.